src/SpectatorView.Native/SpectatorView.Compositor/Compositor/AzureKinectCameraFrame.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "AzureKinectCameraFrame.h"
+
+AzureKinectCameraFrame::AzureKinectCameraFrame(bool captureDepth, bool captureBodyMask)
+    : _captureDepth(captureDepth)
+    , _captureBodyMask(captureBodyMask)
+    , _status(FrameStatus::Unused)
+{
+    _imageSizes[(int)AzureKinectImageType::Color] = FRAME_BUFSIZE_RGBA;
+    _imageSizes[(int)AzureKinectImageType::Depth] = FRAME_BUFSIZE_DEPTH16;
+    _imageSizes[(int)AzureKinectImageType::BodyMask] = FRAME_BUFSIZE_DEPTH16;
+
+    for (int i = 0; i < AZURE_KINECT_IMAGE_TYPE_COUNT; i++)
+    {
+        _images[i] = new uint8_t[_imageSizes[i]];
+    }
+}
+
+AzureKinectCameraFrame::~AzureKinectCameraFrame()
+{
+    for (int i = 0; i < AZURE_KINECT_IMAGE_TYPE_COUNT; i++)
+    {
+        delete[] _images[i];
+    }
+}
+
+void AzureKinectCameraFrame::StageImage(AzureKinectImageType imageType, k4a_image_t image)
+{
+    auto stride = k4a_image_get_stride_bytes(image);
+    auto buffer = k4a_image_get_buffer(image);
+    rsize_t height = k4a_image_get_height_pixels(image);
+
+    _imageStrides[(int)imageType] = stride;
+    memcpy_s(_images[(int)imageType], _imageSizes[(int)imageType], buffer, height * stride);
+}
+
+void AzureKinectCameraFrame::UpdateSRV(AzureKinectImageType imageType, ID3D11Device* device, ID3D11ShaderResourceView* targetView)
+{
+    if (targetView != nullptr)
+    {
+        DirectXHelper::UpdateSRV(device, targetView, _images[(int)imageType], _imageStrides[(int)imageType]);
+    }
+}
+
+bool AzureKinectCameraFrame::TryBeginWritingColorAndDepth()
+{
+    std::lock_guard<std::mutex> guard(_statusGuard);
+    if (_status == FrameStatus::WritingColorAndDepth)
+    {
+        return true;
+    }
+    else if (_status == FrameStatus::Unused || _status == FrameStatus::Staged)
+    {
+        _status = FrameStatus::WritingColorAndDepth;
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+void AzureKinectCameraFrame::EndWritingColorAndDepth()
+{
+    std::lock_guard<std::mutex> guard(_statusGuard);
+    if (_status == FrameStatus::WritingColorAndDepth)
+    {
+        _status = FrameStatus::Staged;
+    }
+}
+
+void AzureKinectCameraFrame::BeginWritingBodyMask()
```

Comment:
>BeginWritingBodyMask [](start = 29, length = 20)

is there a reason that we have separate status for writing body mask vs writing color and depth?

Reply:
It seems like we always write body mask while already in the color and depth mode, so there may not be a reason to have this additional status?

---
In reply to: [403178909](https://github.com/microsoft/MixedReality-SpectatorView/pull/356#discussion_r403178909) [](ancestors = 403178909)